src/Avalonia.Base/Input/DataTransferItem.cs

```diff
@@ -0,0 +1,193 @@
+
+        return null;
+    }
+
+    /// <summary>
+    /// Sets the value for a given format.
+    /// </summary>
+    /// <param name="format">The format.</param>
+    /// <param name="value">
+    /// The value corresponding to <paramref name="format"/>.
+    /// If null, the format won't be part of the <see cref="DataTransferItem"/>.
+    /// </param>
+    public void Set<T>(DataFormat format, T value)
+    {
+        ThrowHelper.ThrowIfNull(format);
+
+        if (value is null)
+            RemoveCore(format);
+        else
+            SetCore(format, new DataAccessor(static state => state, value));
+    }
+
+    /// <summary>
+    /// Sets a value created on demand for a given format.
+    /// </summary>
+    /// <typeparam name="T">The value type.</typeparam>
+    /// <param name="format">The format.</param>
+    /// <param name="getValue">A function returning the value corresponding to <paramref name="format"/>.</param>
+    public void Set<T>(DataFormat format, Func<T> getValue)
+    {
+        ThrowHelper.ThrowIfNull(format);
+        ThrowHelper.ThrowIfNull(getValue);
+
+        SetCore(format, new DataAccessor(static state => ((Func<T>)state)(), getValue));
+    }
+
+    private void SetCore(DataFormat format, DataAccessor accessor)
+    {
+        if (_accessorByFormat is not null)
+            _accessorByFormat[format] = accessor;
+        else if (_singleItem is { } singleItem && !singleItem.Key.Equals(format))
+        {
+            _accessorByFormat = new()
+            {
+                [singleItem.Key] = singleItem.Value,
+                [format] = accessor
+            };
+            _singleItem = null;
+        }
+        else
+            _singleItem = new(format, accessor);
+
+        _formats = null;
+    }
+
+    private void RemoveCore(DataFormat format)
+    {
+        bool removed;
+
+        if (_accessorByFormat is not null)
+            removed = _accessorByFormat.Remove(format);
+        else if (_singleItem is { } singleItem && singleItem.Key.Equals(format))
+        {
+            _singleItem = null;
+            removed = true;
+        }
+        else
+            removed = false;
+
+        if (removed)
+            _formats = null;
+    }
+
+
+    /// <summary>
+    /// Sets the value for the <see cref="DataFormat.Text"/> format.
+    /// </summary>
+    /// <param name="value">
+    /// The value corresponding to the <see cref="DataFormat.Text"/> format.
+    /// If null, the format won't be part of the <see cref="DataTransferItem"/>.
+    /// </param>
+    public void SetText(string? value)
+        => Set(DataFormat.Text, value);
+
+    /// <summary>
+    /// Sets the value for the <see cref="DataFormat.File"/> format.
+    /// </summary>
+    /// <param name="value">
+    /// The value corresponding to the <see cref="DataFormat.File"/> format.
+    /// If null, the format won't be part of the <see cref="DataTransferItem"/>.
+    /// </param>
+    public void SetFile(IStorageItem? value)
```

Comment:
IStorageItem can be a folder, "File" is not perfect here, though we got it since before